packages/world/ts/callWithSignatureTypes.ts

```diff
@@ -0,0 +1,10 @@
+// Follows https://viem.sh/docs/actions/wallet/signTypedData#usage
+
+export const callWithSignatureTypes = {
+  Call: [
+    { name: "delegator", type: "address" },
+    { name: "systemId", type: "bytes32" },
+    { name: "callData", type: "bytes" },
+    { name: "nonce", type: "uint256" },
+  ],
+} as const;
```

Comment:
I wonder if there's a way we can express this as a struct in solidity then import the ABI here instead of manually defining in both spots

Reply:
I agree this would be great, I'm surprised I haven't seen it anyone do it. The problem is we need to derive the EIP712 typehash:
```solidity
bytes32 constant CALL_TYPEHASH = keccak256("Call(address delegator,bytes32 systemId,bytes callData,uint256 nonce)");
```

We could define a struct:

```solidity
struct Call {
  address delegator,
  bytes32 systemId,
  bytes memory callData,
  uint256 nonce
}
``` 

But we'd have to somehow get the names and datatypes of each member as strings, then concat them. It's like we need `.selector` for structs!

Maybe we could define a function selector instead? 🤔  I'll keep thinking about it

Reply:
I see what you mean, see we define it twice, but in the same place? As struct and hash.

How do you include a struct in the ABI though? also we'd need to transform the ABI item into the proper viem type

Reply:
ahh I think I assumed structs would be part of the ABI but they aren't, unless they're used as a function arg but even then it's an ABI for the function args with an `internalType` named `struct Call` (not just `Call`)

close but not quite, but maybe close enough?